SCRUM-6-Installation-und-Konfiguration-von-Drizzle-ORM

calipee opened the pull request:
Ich habe Drizzle installiert und hinzugefügt. Als Schema habe ich direkt das mit @RemodifyBerlin abgesprochene verwendet. 

Ich habe die Erstellung auch lokal getestet, in dem ich einfach eine einfache Seite in Nextjs erstellt habe und so einen Lehrer erstellt habe.

Ich habe es nicht geschafft dafür Tests in Jest zu schreiben und auch online nichts aussagekräftiges dazu gefunden. Hat damit einer von euch Erfahrung @DanielEdlich @RemodifyBerlin ?

Ich bekomme immer einen Fehler wegen fehlendem Fetch. 


SCRUM-6
SCRUM-19

calipee committed:
Add drizzle configuration

SCRUM-6
SCRUM-19

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -0,0 +1,94 @@
+  last_name: text('last_name').notNull(),    // Last name
+  email: text('email').notNull(),            // Email
+  phone_number: text('phone_number'),        // Phone number
+  avatar: text('avatar'),                    // Avatar
```
Hi, Avatar brauchen wir nicht als Attribut. Ich habe ein Component in der Lehrer-Ansicht, die ein Avatar aus den ersten Buchstaben des Namens automatisch bildet. Danke trotzdem!

calipee replied:
Alles klar, werde ich entfernen!

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -0,0 +1,94 @@
+  email: text('email').notNull(),            // Email
+  phone_number: text('phone_number'),        // Phone number
+  avatar: text('avatar'),                    // Avatar
+  priority: integer('priority')              // Priority
```
Macht es Sinn, die Priorität als ein float zu machen? Hier ein Use-Case: es gibt 10 Lehrer mit der Priorität 1. Nun hat der eine Lehrer doch mehr Priorität bekommen - dann muss man bei 9 Lehrern die Priorität zu 2 ändern. Bei einem Float könnte man dann einfach es auf 0.9 wechseln. Wie sieht ihr es?

calipee replied:
Ich denke nicht das wir es als Float machen sollten. Ich finde man muss sich irgendwie einschränken wie die Priorisierung ist sonst wird das irgendwie unschön weil dann ganz schnell immer kleinerer Abstände genommen werden.

Alternativ könnten wir die Priorisierung auch irgendwie dynamisch machen, in dem man die Lehrer einfach anordnen bzw auf eine Ebene setzen kann und daraus im Hintergrund die Priorisierung berechnet wird.

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -0,0 +1,94 @@
+// ---------------------------------------
+export const teacherAvailability = sqliteTable('teacher_availability', {
+  id: integer('id', { mode: 'number' }).primaryKey({ autoIncrement: true }),
+  day: text('day').notNull(),                // Day (e.g. Monday, Tuesday...)
```
Geht das als Enum?

calipee replied:
Es gibt keine Enums in SQLite, aber wir können das dann in den Modelklassen als Enum definieren.

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -0,0 +1,94 @@
+export const schoolClass = sqliteTable('school_class', {
+  id: integer('id', { mode: 'number' }).primaryKey({ autoIncrement: true }),
+  name: text('name').notNull(),      // Class name
+  year: integer('year'),             // Year
```
Hier vielleicht eher ein String? Denn es sind Klassen von mehreren Jahrgängen. zB Klasse "1-3" oder "4-6"

calipee replied:
Guter Vorschlag, daran hatte ich garnicht gedacht.

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -0,0 +1,94 @@
+export const lesson = sqliteTable('lesson', {
+  id: integer('id', { mode: 'number' }).primaryKey({ autoIncrement: true }),
+  date: text('date'),                         // Date (ISO-String, e.g. "2024-12-06")
+  timeslot_from: integer('timeslot_from').notNull(), // Timeslot start (e.g. 8)
```
Hier macht ein einzelnes Zeitfenster mehr Sinn. Es handelt sich immer um eine Unterrichtseinheit.

calipee replied:
Okay, wir sollten uns dann noch überlegen wie wir mit Fächern umgehen die in Blöcken unterrichtet werden

RemodifyBerlin replied:
Denke es macht Sinn eine Klon-Funktion bei einen Unterrichtsblock bzw. Kärtchen zu haben. So kann man einfach ein Unterricht klonen und man hat schon ein block. 

Alternativ ein Boolean: is_block

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -0,0 +1,94 @@
+// ---------------------------------------
+export const lesson = sqliteTable('lesson', {
+  id: integer('id', { mode: 'number' }).primaryKey({ autoIncrement: true }),
+  date: text('date'),                         // Date (ISO-String, e.g. "2024-12-06")
```
Date macht hier nicht Sinn, denn es ist einfach ein Raster von 2 Wochen, ohne bestimmte Daten. Hier lieber ein Wochentag-Enum wie bei den Lehrer-Verfügbarkeiten

calipee replied:
Ja, du hast völlig Recht. Sollten wir dann irgendwie noch A/B Woche mit aufnehmen?

calipee replied:
Ja, du hast völlig Recht. Sollten wir dann irgendwie noch A/B Woche mit aufnehmen?

RemodifyBerlin replied:
Genau ja, gute Idee

calipee committed:
Refactor: Update schema

- Change Schoolclass year to text because classes are in years 1-3 and 4-6
- change from and to timeslot in lesson to single timeslot because one lesson is always one timeslot
- remove field for avatar which will not be needed because avatars will be generated based on the teachers name

RemodifyBerlin commented on src/db/schema.ts:
```diff
@@ -60,10 +59,9 @@ export const timetable = sqliteTable('timetable', {
-  timeslot_from: integer('timeslot_from').notNull(), // Timeslot start (e.g. 8)
-  timeslot_to: integer('timeslot_to').notNull(),     // Timeslot end (e.g. 10)
-  lesson_name: text('lesson_name'),     // Lesson content
+  date: text('date'),                         // Weekday as Enum (Monday, Tuesday, etc)
```
Mir ist auch aufgefallen das es A und B wochen gibt. Entweder nennen wir die Tage nach dem Schema Montag_A oder wir müssen noch ein Feld ergänzen: Woche ("A" oder "B")

calipee committed:
Add week field to lesson schema for A and B weeks

calipee committed:
Merge remote-tracking branch 'origin/main' into SCRUM-6-Installation-und-Konfiguration-von-Drizzle-ORM
